Composer/packages/adaptive-form/src/components/fields/IntellisenseFields.tsx

```diff
@@ -0,0 +1,173 @@
+      value={value}
+      onBlur={props.onBlur}
+      onChange={onChange}
+    >
+      {({ textFieldValue, focused, onValueChanged, onKeyDownTextField, onKeyUpTextField, onClickTextField }) => (
+        <StringField
+          {...props}
+          focused={focused}
+          id={id}
+          value={textFieldValue}
+          onBlur={undefined} // onBlur managed by Intellisense
+          onChange={(newValue) => onValueChanged(newValue || '')}
+          onClick={onClickTextField}
+          onKeyDown={onKeyDownTextField}
+          onKeyUp={onKeyUpTextField}
+        />
+      )}
+    </Intellisense>
+  );
+};
+
+export const IntellisenseNumberField: React.FC<FieldProps<string>> = function IntellisenseNumberField(props) {
+  const { id, value = '', onChange, uiOptions, focused: defaultFocused } = props;
+
+  const completionListOverrideResolver = (value: string) => {
+    if (value === '') {
+      return <ExpressionSwitchWindow type={'Number'} onSwitchToExpression={() => onChange('=')} />;
+    } else {
+      return null;
+    }
+  };
+
+  return (
+    <Intellisense
+      completionListOverrideResolver={completionListOverrideResolver}
+      focused={defaultFocused}
+      id={`intellisense-${id}`}
+      scopes={uiOptions.intellisenseScopes || []}
+      url={getIntellisenseUrl()}
+      value={value}
+      onBlur={props.onBlur}
+      onChange={onChange}
+    >
+      {({ textFieldValue, focused, onValueChanged, onKeyDownTextField, onKeyUpTextField, onClickTextField }) => (
+        <NumberField
+          {...props}
+          focused={focused}
+          id={id}
+          value={textFieldValue}
+          onBlur={undefined} // onBlur managed by Intellisense
+          onChange={(newValue) => onValueChanged(newValue || 0)}
+          onClick={onClickTextField}
+          onKeyDown={onKeyDownTextField}
+          onKeyUp={onKeyUpTextField}
+        />
+      )}
+    </Intellisense>
+  );
+};
+
+export const IntellisenseJSONField: React.FC<FieldProps<string>> = function IntellisenseJSONField(props) {
+  const { id, value = '', onChange, focused: defaultFocused, schema } = props;
+
+  const completionListOverrideResolver = (value: any) => {
+    if (typeof value === 'object' && Object.keys(value).length === 0) {
+      return (
+        <ExpressionSwitchWindow
+          type="Object"
+          onSwitchToExpression={() => {
+            onChange('=');
+          }}
+        />
+      );
+    } else if (Array.isArray(value) && value.length === 0) {
+      return (
+        <ExpressionSwitchWindow
+          type="Array"
+          onSwitchToExpression={() => {
+            onChange('=');
+          }}
+        />
+      );
+    }
+    return null;
+  };
+
+  const defaultValue = schema.type === 'object' ? {} : [];
+
+  return (
+    <Intellisense
+      completionListOverrideResolver={completionListOverrideResolver}
+      focused={defaultFocused}
+      id={`intellisense-${id}`}
+      scopes={['expressions']}
```

Comment:
This allocates a new array every single render which will trigger a any effects on scopes to fire within Intellisense..  Put constants into const variables outside render.

Reply:
Okay!